### ISSUE TITLE
Please, add rule: (either Left) == (=<<)

### ISSUE THREAD

**Issue by Anton-Latukha** [ndmitchell/hlint#1308]
Noticed that `=<<` needs to be recommended there.

Generally, all these kinds of `switcher mzero f x` Monad functions that does not change the outside Monad - reduce into `=<<`, since:
1. Canonical Lambda argument order for binary control path `switcher`-type function is to pass case for `mzero` first.
2. In case of that being same-monad `mzero` that gets analyzed (`x`, `Left` in this case) - it just reconstructs the `mzero` of `x`.
3. `mzero` literal defining law/property is that it short-circuits `=<<`. So `(f =<< mzero) == mzero`, which inside the context of `Either` `Monad` `Left`s seems equivalent to `either Left f (Left a)`.
4. Doing typecheck:

```haskell
-- ghci>
  
  :t (either Left)
    
    (either Left)
      :: (b1 -> Either a b2) -> Either a b1 -> Either a b2
    
    
  :t ((=<<) :: (b1 -> Either a b2) -> Either a b1 -> Either a b2)
    
    ((=<<) :: (b1 -> Either a b2) -> Either a b1 -> Either a b2)
      :: (b1 -> Either a b2) -> Either a b1 -> Either a b2
        
-- the type of (either Left f x) equal to (f =<< x), in the context of Either monad, which was needed to be proven.
```

5. I am not proficient in Hask bottom to elaborate on them.
6. `f =<< x` makes code from concrete `Either` into polymorphic `Monad`, which is generally regarded as a good thing, but, just as `maybe Nothing -> =<<`, with `x` changing the type would act accordingly to literal semantics of code.
7. The separation of concerns is one of the main code paradigms, the main separation of concerns in Haskell language design is the separation of code into 3 code layers: {term, type, kind}: so types are in the type code. `either` is a term-level function that enforces concrete type, which by comparison between `(either Left) <=> (=<<)` in `either` is the violation of separation on concerns, as it enforces the concrete type inside the term level implicitly, & that approach requires a new function for every type to enforce it so it is unsustainable language design & approach (on the big scale): if to consider the memoization & use of mapping `(endless set of concrete types -> requires even faster-growing endless set of function sets)`.
8. Usage of general functions as `=<<` promotes the use of type-level code to control types, & `=<<` is the main path of code refactoring/optimization that enables further recursion on `hlint` rules, often resulting in major code simplifications `=<< -> {monad laws, <$> (& functor laws), reduction of "do", Traversable (& so Traversable laws, ...}`.

**Comment by ndmitchell** [comment]
I think the simpler argument is that we already do this for `- hint: {lhs: maybe Nothing f x, rhs: f =<< x}` so adding it for `either` is just following the same principles there. As a general principle, we don't always aim to make things more polymorphic (there is a subset of Haskell programmers who consider everything polymorphic to be a good thing, but its not universal), which is why we have the generalise package of hints. That said, this one seems a good idea anyway.
